### PR TITLE
Zero initialize static data section only in the main thread and add test for global static objects with dylink + pthreads.

### LINKED ISSUE #13386: Global objects in dynamic libraries get reallocated for each thread
When a worker for a thread is spawned, the heap space for the dynamic libraries gets reallocated instead of reusing the existing heap.
This makes global objects in dynamic libraries behave more like thread local than global shared.

side.cpp
```c++
int value = 42;
int & get_value() {
    return value;
}
```

main.cpp
```c++
#include <stdio.h>
#include <thread>
int & get_value();
int main(void) {
    auto t1 = std::thread([]{
      get_value() = 1;
      printf("%d ", get_value());
    });
    std::thread([&]{
      // make sure this thread doesn't reuse
      // the worker from the previous one.
      t1.join();
      printf("%d\n", get_value());
    }).join();
    return 0;
}
```

compile commands:
```
em++ -o side.wasm side.cpp -pthread -Wno-experimental -s SIDE_MODULE=1
em++ -pthread -Wno-experimental -s PROXY_TO_PTHREAD -s EXIT_RUNTIME=1 -s RUNTIME_LINKED_LIBS=['side.wasm'] -s MAIN_MODULE=1
```

Expected output
```
1 1
```

Actual output
```
1 42
```

### PR DESCRIPTION
**Pull request by jprendes**
Along with #13603, this change fixes #13386

With #13603, the static data section for shared libraries is shared among different threads.
It also happened to fix global constructors being called in threads (additionally to the main thread).

This change fixes the remaining issue of the static data section being zero-initialized in each thread (additionally to the main thread). Fixing this remaining issue #13386 should be fixed.

This change also adds a test to ensure global static objects in shared libraries are handled correctly.

### CODE REVIEW

**Comment by jprendes** [comment]
CI check `test-mac` didn't run because of my circleci's plan.

File: src/library_dylink.js
Comment (sbc100) [inline src/library_dylink.js:416]: Is this assuming that the main thread is also the first thread to load the module?    I'm not sure we have decided this can/is always the case.    we might instead need some shared data structure (in shared memory) that lists all the module along with their memory base, that way when we come to load a give module we know if we are first and we can allocate and initialize the memory, otherwise we can skip it.

**Review by sbc100 — COMMENTED** [review]
Thanks again for working on this. 

I'm wondering if we need some kind of new shared data structure to keep track of modules, the order in which they are loaded, and their base addresses.   I was planning to add something like this but havn't got there yet myself.

File: src/library_dylink.js
Comment (jprendes) [reply src/library_dylink.js:416]: The main thread doesn't need  to be the first thread to load, but it needs to load before running code using globals in any thread. AFAIK that holds true, is that right?

The main thread zero-initialize and calls the constructors. The other threads just point their `__heap_base` to the same location as the main thread, and that's why the main thread needs to load before using any globals.

The change does assume that dlopen will not load a new library (all side libraries are pre-loaded). But I think that's a general assumption with pthread + dylink at the moment.

In the long run I agree that having a structure in shared memory is a better solution (it would be more robust and scale better to cases like allowing arbitrary dlopen). I can give it a go if you think that should be included as part of this change, otherwise I am also happy to give it a go as a follow up change.

Comment (sbc100) [reply src/library_dylink.js:416]: In general I'm OK with with that happening as a followup, but I'm not a fan the change to `getMemory` as it stands.   Perhaps if we renamed `getMemory` to something more specific such as `getModuleDataAddress`  or `getOrAllocateModuleMemory` or something like that?   

As it stands `getMemory` has historically been a generic allocator and this change certainly changes that.

File: src/library_dylink.js
Comment (sbc100) [inline src/library_dylink.js:252]: How about moving this check outside the outer one?

Can you add to the comment here something about like "because we only support load-time dynamic linking with pthreads today" and perhaps a TODO about redesigned this to use shared layout once we have runtime dynamic linking?

File: src/library_dylink.js
Comment (sbc100) [reply src/library_dylink.js:416]: Can you use opening and closing braces here and above.. to match out JS style.

**Review by sbc100 — APPROVED** [review]

File: src/library_dylink.js
Comment (jprendes) [reply src/library_dylink.js:416]: On a second thought, the shared memory approach would require some kind of locking/mutex, and that wouldn't play nicely with the main thread.

I'm somewhat confused since `__wasm_init_memory` does wait if someone else is initializing the memory. Is it guaranteed to be the main thread in this case?

An alternative would be to proxy all the operations to the main thread, so that the main thread never needs to wait, but then there's not much to gain from storing the data in shared memory.

Comment (sbc100) [reply src/library_dylink.js:416]: Imagine you load a module at runtime, you need to call malloc to allocate its memory.  Now the same module needs to be loaded on every other thread, and the base address (__memory_base) needs to be the same as it was on the main thread.  So at the very least we need to store the base address of each module in shared memory (or somehow share it between threads).  

__wasm_init_memory works I believe because all it cares about is that it runs exacltly.  Its results don't need to be shared between theads in the way the getMemory results do (IIUC).

Comment (jprendes) [reply src/library_dylink.js:418]: My confusion about `__wasm_init_memory` is about the waiting, since a shared memory solution here could use a similar strategy to ensure only one thread is writing to it. Where does the guarantee that the wait doesn't happen in the main thread come from?

The case I am thinking about is the main thread trying to load a library while another thread is already allocating the memory for it. The main thread would have to wait until the other thread finishes allocating (and initializing) that memory so that it can use the same address. This wait operation might not always be possible (e.g., browser render thread). I believe this is the same challenge faced by `__wasm_init_memory`, but I don't understand why it works in that case.

Is this PR the right place for this discussion? Or should I move this to a different forum?

Comment (sbc100) [reply src/library_dylink.js:418]: I think the reason `__wasm_init_memory` never blocks on the main thread today is simply that the main thread always instantiates the module before postMessaging it to background threads, so it always wins the race.

I think that the synchronization of the various module across all the threads is the single biggest challenge with pthreads + dynamic linking.   There are many unanswered questions and a lot of them relate to how we do the loading (and potentially the blocking) on different threads.   Perhaps the best place to discuss these kinds of issues is on #3494?

**Comment by sbc100** [comment]
I hope you don't mind by I found a different solution to the getMemory issue:  https://github.com/emscripten-core/emscripten/pull/13603.

I think something more advanced will be needed either way once we get to runtime linking, but I prefer the less intrusive approach in #13603 for now I think.   

The other parts of this change should still be good to go.

**Comment by jprendes** [comment]
> I hope you don't mind by I found a different solution to the getMemory issue: #13603.
> 
> I think something more advanced will be needed either way once we get to runtime linking, but I prefer the less intrusive approach in #13603 for now I think.

I agree, #13603 is a better solution. Still wouldn't be enough for runtime `dlopen`, but that's a different can of worms.

> The other parts of this change should still be good to go.

I'll revert the `getMemory` parts. Then #13603 would need to be merged before this change.

**Comment by jprendes** [comment]
I will merge and push again once #13603 is merged.
Doesn't make much sense to run the CI checks until then.

File: tests/test_core.py
Comment (jprendes) [inline tests/test_core.py:8305]: This one puzzles me, I don't know why this particular test function didn't get updated in my fork when I clearly see it updated in the github source.

Comment (sbc100) [reply tests/test_core.py:8305]: Sorry, that a bad commit on my part.  Should be fixed now.

File: tests/test_other.py
Comment (sbc100) [inline tests/test_other.py:1552]: How about calling this `get_address()` and returning a pointer .. it seems a little more descirptive.

File: tests/test_other.py
Comment (sbc100) [inline tests/test_other.py:1528]: The term heap is a littel misleading here I think.   How about refering to it as he static data region (or segment) of of the side module?

Perhaps this test can be called `test_dynamic_link_pthread_static_data`?

File: tests/test_other.py
Comment (sbc100) [inline tests/test_other.py:1537]: Can this just be replaced with `static int data = 42`?

**Review by sbc100 — COMMENTED** [review]
This is looking good now I think,  just a couple of nits regarding the test.

File: tests/test_other.py
Comment (jprendes) [reply tests/test_other.py:1537]: A static int would be initialized when the memory segment gets initialized, which happens during module instantiation.
An object with a constructor would get initialized when global constructors are called, which is triggered by `__post_instantiate`.
One of the fixes in this change is that `__post_instantiate` is only called once (in the main thread). Using an object with a constructor was a deliberate choice.

Comment (sbc100) [reply tests/test_other.py:1539]: Ah I see good point.

How would this test detect if the constructor was run twice?

Perhaps we should have the constructor have side effect other than data initilization so that it can't be eliminated completely by an optimizer?  Perhaps it could write do a `+= 1` on the `static int`?    It might also make it easier to debug if it included a `puts("ctor called")` so that you can see in the console if its called twice?

Comment (jprendes) [reply tests/test_other.py:1539]: Ah, good call.
When I wrote this change, this was true:

The constructor initializes the static int to `42`.
The code in `main` sets the value to `123` before creating the thread.
If creating the thread runs the constructor again, it resets the value to `42`.

However, with #13603, constructors are only called once anyway:
Before #13603, `__post_instantiate` was called immediately on the pthread.
After #13603 the `runtimeInitialized` check changed, `__post_instantiate` gets pushed into `__ATINIT__`. However, the content of `__ATINIT__` is not executed in the pthread (`initRuntime` has an early return), resulting in the constructors being called only once.

So I guess what this change really does now is not zero-initialize the static memory segment twice, and avoid a pointless entry being added to `__ATINIT__` in pthreads.

Comment (jprendes) [reply tests/test_other.py:1539]: To recap:
* Removing the `if (!ENVIRONMENT_IS_PTHREAD) {` check that guards the call to `__post_instantiate` has no effect now, other than preventing an irrelevant push to `__ATINIT__`.
* Removing the `if (!ENVIRONMENT_IS_PTHREAD) {` check that guards the zero-init of the memory does have an effect, but that would be true even if the test used a `static int` rather than an object with constructor.

I wonder if there's still value in having a test that verifies the constructor runs once or not. It's a happy but AFAIK untested consequence of #13603.

Comment (sbc100) [reply tests/test_other.py:1539]: Thank you for the explanation.   I think both your fixes are good, I just didn't fully understand what was happening in the test.

Can perhaps add a comment to the test saying what is trying to test for exactly.  Something like:  "Tests that thread loading of dynamic libraries that happen on thread creation does not zero-only data segments or call static constructors".?

Other this I think this is good to go now.

Comment (sbc100) [reply tests/test_other.py:1539]: I see you already added a comment above which seems fine.

File: tests/test_other.py
Comment (sbc100) [inline tests/test_other.py:1530]: regionfor -> region for

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
Thanks for the all the work and updated on the PR!   Can you update the PR title and description to match where we ended up and then I can land it.

**Comment by jprendes** [comment]
Title and description updated!
Thanks for taking the time to review the changes!

**Pull request merged**
